---
title: "Figure_5"
author: "Max Cummins"
date: "`r Sys.Date()`"
output:
    html_document:
        theme: spacelab
        toc: true
        toc_float: true
---

#

```{r setup, include=FALSE}
#Set wd to directory containing current markdown script
knitr::opts_chunk$set(dirname(rstudioapi::getSourceEditorContext()$path))

#Read in our packages
library(tidyverse)
library(ggplot2)
library(xml2)
library(ggtree)
library(paletteer)
library(vroom)
library(ggpubr)


#Define our not in function
`%nin%` <- Negate(`%in%`)
```

## Read in our necessary files

```{r}

#Read in our genometa sheet generated by AusTrakka_TEA_data_generation.Rmd
genometa <- vroom("delims/genometa_n5631.txt", show_col_types = FALSE)

#Read in our tree
tree <- ggtree::read.tree("analysis/pad.nwk")

#Remove quotes from sample names
tree$tip.label <- gsub("'", "", tree$tip.label)

#Create a list of samples in our cohort
sample_names <- tree$tip.label %>% as.data.frame() %>% rename('name' = '.')
```

## Define any variables

```{r}
#Overwrite with different colours
source_cols <- c("Companion Animal" = "#59398d",
         "Environmental" = "#709b46",
         "Food" = "#c09f3d",
         "Human" = "#48c595",
         "Livestock" =  "#c26bbc",
         "Wild Animal" = "#b9553d")
```

## Generate figure
```{r}

df_all_ST = data.frame()

for(i in list.dirs("analysis/SKA")){
        if(str_detect(string = i, pattern = "_dists")){
                #ST extraction
                ST <- gsub("_dists", "", basename(i))
                #Read in SNP distances
                dists <- read_delim(paste0(i,"/",ST,"_dist_100_SNP.distances.tsv"), 
                    delim = "\t", escape_double = FALSE, 
                    trim_ws = TRUE, show_col_types = FALSE)
                
                #Read in SNP Clusters
                clusts <- read_delim(paste0(i,"/",ST,"_dist_100_SNP.clusters.tsv"), 
                    delim = "\t", escape_double = FALSE, 
                    trim_ws = TRUE, show_col_types = FALSE)
                
                
                #Combine our metadata and our clusters for use in microreact
                data_microreact <- left_join(clusts, genometa, by = c("ID" ="name")) %>% select(ID, starts_with("Revised_Source"), Collection_Year, Collection, fimtyper..Fimtype, ColV, Cluster__autocolour)
                
                data_microreact1 <- data_microreact %>% rename("Sample 1" = "ID") %>% rev()
                
                data_microreact2 <- data_microreact %>% rename("Sample 2" = "ID")
                
                dists2 <- inner_join(data_microreact1, dists, by = "Sample 1")
                
                dists2 <- inner_join(dists2, data_microreact2, by = "Sample 2")
                
                dists2 <- dists2 %>% mutate(across(where(is.factor), as.character))
                
                dists_w_metadata <- dists2 %>% rowwise() %>%
                        mutate(col1 = pmin(Revised_Source_Niche.x, Revised_Source_Niche.y), col2 = pmax(Revised_Source_Niche.x, Revised_Source_Niche.y), source_combo = paste0(col1, "/",col2, collapse = "")) %>% select(-col1, -col2)
                
                dists_w_metadata <- dists_w_metadata %>% rowwise() %>%
                        mutate(col1 = pmin(Revised_Source_Type.x, Revised_Source_Type.y), col2 = pmax(Revised_Source_Type.x, Revised_Source_Type.y), source_type_combo = paste0(col1, "/",col2, collapse = "")) %>% select(-col1, -col2)
                
                dists_w_metadata$ST.x <- ST
                
                df_all_ST <- rbind(df_all_ST, dists_w_metadata)
                
               # write_delim(x = data_microreact1, file = paste0(i,"/",ST,"_metadata_microreact.tsv"), delim = "\t")
        }
        
}
```

```{r}
#library(harrietr)

df_all_ST <- df_all_ST %>% as.data.frame() %>% filter(`Sample 1` %in% tree$tip.label & `Sample 2` %in% tree$tip.label)

dists_w_metadata_x_source <- df_all_ST %>% as.data.frame() %>% filter(Revised_Source_Niche.x != Revised_Source_Niche.y)

cgMLST_clusters <- read_delim(file = "analysis/clusters.txt")

allelic_distances <- read_delim(file = "analysis/pad.txt", show_col_types = FALSE)

allelic_distances <- allelic_distances %>% rename(name = '5631')

long_dists <- pivot_longer(allelic_distances, cols = 2:ncol(allelic_distances), names_repair = 'minimal')

colnames(long_dists) <- c('Sample 1', 'Sample 2', 'value')

distsnew <- inner_join(df_all_ST, long_dists)

dists_w_metadata_all_source <- distsnew %>% as.data.frame() 

df_all_ST_filtered_SNPs_all <- dists_w_metadata_all_source %>% as.data.frame() 

pal <- colorRampPalette(c("red", "yellow", "blue"))
dot_cols <- pal(41)

df_all_ST_filtered_SNPs_all %>%
        mutate(source_combo = str_replace(source_combo, 'Companion Animal', 'Comp. Animal')) %>% 
        mutate(source_combo = str_replace(source_combo, 'Environmental', 'Environ.')) %>% 
        filter(Revised_Source_Niche.x != Revised_Source_Niche.y) %>% 
        filter(value < 40)  %>% 
        mutate(value = as.factor(value)) %>%
        rename('Allelic\nDistance\n(cgMLST)' = value) %>%
        rename('Sequence type' = ST.x) %>%
        ggplot() +
        geom_boxplot(aes(x = SNPs, y = `Sequence type`)) +
        geom_jitter(aes(x = SNPs, y = `Sequence type`, color = `Allelic\nDistance\n(cgMLST)`), size = 0.25) +
        scale_color_manual(values = dot_cols, breaks = as.factor(0:40)) +
        facet_wrap(~source_combo, nrow = 7, ncol = 2, strip.position="right") +
        theme(axis.text.x = element_text(angle = 90)) +
        geom_vline(xintercept = 100, color = "red", linetype = "dashed") +
        scale_x_continuous(breaks = seq(0,1300, by = 100)) +
        guides(colour = guide_legend(override.aes = list(size=1.5), ncol = 1))
```

Supplementary Figure X
```{r}
plot1 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 12000) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = same_fim)) +
        facet_wrap(~ST.x) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label")# +

ggsave(filename = "/Users/131785/Library/CloudStorage/OneDrive-TheUniversityofMelbourne/Figures/SNPs_vs_cgMLST_facet_ST_w_fim.pdf", plot = plot1, width = 11.69, height = 8.27, units = "in", dpi = 300)
```

Supplementary Figure X
```{r}

STs <- df_all_ST_filtered_SNPs_all$ST.x %>% unique()

ST_cols <- paletteer_d("ggthemes::Tableau_10") %>% as.vector()

names(ST_cols) <- STs


plot2 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 12000) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = ST.x), alpha = I(0.2)) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label") +
        scale_color_manual(values = ST_cols)

ggsave(filename = "/Users/131785/Library/CloudStorage/OneDrive-TheUniversityofMelbourne/Figures/SNPs_vs_cgMLST_col_by_ST.pdf", plot = plot2, width = 11.69, height = 8.27, units = "in", dpi = 300)

```
```{r}
plot3 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 100) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = ST.x), alpha = I(0.4)) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label") +
        scale_color_manual(values = ST_cols)

ggsave(filename = "/Users/131785/Library/CloudStorage/OneDrive-TheUniversityofMelbourne/Figures/SNPs_vs_cgMLST_col_by_ST_sub_100.pdf", plot = plot3, width = 11.69, height = 8.27, units = "in", dpi = 300)

plot4 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 100) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_fim = if_else(fimtyper..Fimtype.x == fimtyper..Fimtype.y, T, F)) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = same_fim), alpha = I(0.2)) +
        facet_wrap(~ST.x) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman", aes(label = paste(..rr.label.., ..p.label.., sep = "~")), color = "red", geom = "label")

ggsave(filename = "/Users/131785/Library/CloudStorage/OneDrive-TheUniversityofMelbourne/Figures/SNPs_vs_cgMLST_facet_ST_w_fim_sub_100.pdf", plot = plot4, width = 11.69, height = 8.27, units = "in", dpi = 300)
```

```{r}
plot5 <- df_all_ST_filtered_SNPs_all %>% filter(SNPs < 12000) %>%
        mutate(fimtyper..Fimtype.x = if_else(is.na(fimtyper..Fimtype.x), "Missing", fimtyper..Fimtype.x)) %>%
        mutate(fimtyper..Fimtype.y = if_else(is.na(fimtyper..Fimtype.y), "Missing", fimtyper..Fimtype.y)) %>%
        mutate(same_source = if_else(Revised_Source_Niche.x == Revised_Source_Niche.y, "Intersource", "Intrasource")) %>% 
        ggplot(aes(x = SNPs, y = value)) + #, color=intra_inter
        geom_point(aes(color = ST.x), alpha = I(0.4)) +
        facet_wrap(~same_source) +
        labs(x = "Pairwise SNP Distance", y = "Pairwise Allelic Distance") +
        geom_smooth(
                method = "glm",
                color = "red",
                fill = "#69b3a2",
                se = TRUE
        ) +
        stat_cor(method = "spearman") +
        scale_color_manual(values = ST_cols)

ggsave(filename = "/Users/131785/Library/CloudStorage/OneDrive-TheUniversityofMelbourne/Figures/SNPs_vs_cgMLST_facet_same_source_col_by_ST.pdf", plot = plot5, width = 11.69, height = 8.27, units = "in", dpi = 300)
```

