library(vroom)
library(tidyverse)
library(igraph)

setwd(dir = "APG-OHEC-Retro-M1/markdown/")

distances <- read_delim(file = "../analysis/SKA/ST95_dists/ST95_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST95")

distances2 <- distances

distances <- read_delim(file = "../analysis/SKA/ST131_dists/ST131_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST131")

distances2 <- bind_rows(distances2, distances)

distances <- read_delim(file = "../analysis/SKA/ST117_dists/ST117_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST117")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST963_dists/ST963_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST963")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST69_dists/ST69_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST69")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST457_dists/ST457_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST457")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST57_dists/ST57_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST57")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST648_dists/ST648_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST648")

distances2 <- bind_rows(distances2, distances)

distances <- read_delim(file = "../analysis/SKA/ST1193_dists/ST1193_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST1193")

distances2 <- bind_rows(distances2, distances)

distances <- read_delim(file = "../analysis/SKA/ST80_dists/ST80_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST80")

distances2 <- bind_rows(distances2, distances)

dists <- distances2 %>% select(`Sample 1`, `Sample 2`, SNPs)

#Read in our genometa sheet generated by AusTrakka_TEA_data_generation.Rmd
genometa <- vroom("../delims/genometa_n5631.txt", show_col_types = FALSE)

sequence_type <- "ST95"

set.seed(1)
dists <- distances2 %>%
  filter(ST == sequence_type) %>%
  filter(SNPs <= 100) %>%
  select(`Sample 1`, `Sample 2`, SNPs)

#Extract sources from genotypic/metadata table
sources <- genometa %>% filter(paste0("ST",ST_new) == sequence_type) %>% select(name, matches("Source"))

#Set a colour for each source
sources <- sources %>%
  mutate(color = case_when(
    Revised_Source_Niche == "Captive Animal" ~ "#b5467395",
    Revised_Source_Niche == "Companion Animal" ~ "#59398d95",
    Revised_Source_Niche == "Environmental" ~ "#709b4695",
    Revised_Source_Niche == "Food" ~ "#c09f3d95",
    Revised_Source_Niche == "Human" ~ "#48c59595",
    Revised_Source_Niche == "Livestock" ~ "#c26bbc95",
    Revised_Source_Niche == "Waste" ~ "#6d83da95",
    Revised_Source_Niche == "Wild Animal" ~ "#b9553d95"
  ))

# Convert data frame to edge list for igraph
g <- graph_from_data_frame(d = dists, directed = FALSE)

#Get a list of vertex names
g_vertex_names <- V(g)$name 

# Match colors to vertices based on names
vertex_color_vector <- sources$color[match(g_vertex_names, sources$name)]

# Plot the graph
pdf(file = paste0("../figures/Network_graphs/Single_",
                  sequence_type,
                  ".pdf"),
    width = 11.69,
    height = 8.27)

E(g)$thickness <- ifelse(E(g)$SNPs <= 20, 3,
                         ifelse(E(g)$SNPs <= 50, 2, 1))


set.seed(1)

plot(g, edge.width = E(g)$thickness,
     rescale = TRUE,
     main = sequence_type,
     vertex.size = 3,
     vertex.color = vertex_color_vector,
)

legend("bottomleft",
       legend = c("Captive Animal",
                  "Companion Animal", 
                  "Environmental", 
                  "Food", 
                  "Human", 
                  "Livestock", 
                  "Waste", 
                  "Wild Animal"),
       fill = c(
         "#b5467395",
         "#59398d95",
         "#709b4695",
         "#c09f3d95",
         "#48c59595",
         "#c26bbc95",
         "#6d83da95",
         "#b9553d95"),
       pch=19,
       bty = "n",
       pt.cex = 1,
       cex = 1,
       text.col="black" ,
       horiz = FALSE)

dev.off()



# Plot the graph
pdf(file = paste0("../figures/Network_graphs/Communities_", sequence_type, ".pdf"), width = 11.69, height = 8.27)


clusters <- cluster_walktrap(g)

plot(clusters, g, edge.width = E(g)$thickness,
     rescale = TRUE,
     main = sequence_type,
     vertex.size = 3
)
dev.off()


test <- plot_dendrogram(
  clusters,
  mode = igraph_opt("dend.plot.type"),
  use.modularity = FALSE,
  palette = categorical_pal(8)
)

clusters_w_metadata <- membership(clusters) %>%
  as.data.frame() %>%
  mutate(x = as.integer(x)) %>%
  rownames_to_column('name') %>%
  inner_join(genometa[,1:40], by = 'name')

clusters_w_metadata %>%
  group_by(Revised_Source_Type) %>%
  filter(x == 9) %>%
  summarise(counts = n())

cluster_2 <- clusters_w_metadata %>%
  group_by(Revised_Source_Type) %>%
  filter(x == 9)

#Via running script for Figure 5
df_all_ST_filtered_SNPs_all %>%
  filter(`Sample 1` %in% cluster_2$name & `Sample 2` %in% cluster_2$name) %>%
  filter(Revised_Source_Niche.x == Revised_Source_Niche.y) %>%
  summarise(min = min(SNPs), median = median(SNPs), max = max(SNPs))

#2 = C1
#7 = C2
#8 = C4
#9 = C3
#4 = C5 

