#install.packages("igraph")
library(igraph)

#setwd(dir = "APG-OHEC-Retro-M1/markdown/")

distances <- read_delim(file = "../analysis/SKA/ST95_dists/ST95_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST95")

distances2 <- distances

distances <- read_delim(file = "../analysis/SKA/ST131_dists/ST131_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST131")

distances2 <- bind_rows(distances2, distances)

distances <- read_delim(file = "../analysis/SKA/ST117_dists/ST117_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST117")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST963_dists/ST963_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST963")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST69_dists/ST69_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST69")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST457_dists/ST457_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST457")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST57_dists/ST57_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST57")

distances2 <- bind_rows(distances2, distances)
distances <- read_delim(file = "../analysis/SKA/ST648_dists/ST648_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST648")

distances2 <- bind_rows(distances2, distances)

distances <- read_delim(file = "../analysis/SKA/ST1193_dists/ST1193_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST1193")

distances2 <- bind_rows(distances2, distances)

distances <- read_delim(file = "../analysis/SKA/ST80_dists/ST80_dist_100_SNP.distances.tsv", delim = "\t")

distances <- distances %>% mutate(ST = "ST80")

distances2 <- bind_rows(distances2, distances)

dists <- distances2 %>% select(`Sample 1`, `Sample 2`, SNPs)

#Read in our genometa sheet generated by AusTrakka_TEA_data_generation.Rmd
genometa <- vroom("../delims/genometa_n5631.txt", show_col_types = FALSE)

for(sequence_type in unique(distances2$ST)){

  set.seed(1)
  dists <- distances2 %>%
    filter(ST == sequence_type) %>%
    filter(SNPs <= 100) %>%
    select(`Sample 1`, `Sample 2`, SNPs)
  
  #Extract sources from genotypic/metadata table
  metadata <- genometa %>%
    filter(paste0("ST",ST_new) == sequence_type) %>%
    select(name, Revised_Source_Niche, Collection_Year, State, Collection)

  #Set a colour for each source
  sources <- metadata %>%
    mutate(color = case_when(
      Revised_Source_Niche == "Captive Animal" ~ "#b5467395",
      Revised_Source_Niche == "Companion Animal" ~ "#59398d95",
      Revised_Source_Niche == "Environmental" ~ "#709b4695",
      Revised_Source_Niche == "Food" ~ "#c09f3d95",
      Revised_Source_Niche == "Human" ~ "#48c59595",
      Revised_Source_Niche == "Livestock" ~ "#c26bbc95",
      Revised_Source_Niche == "Waste" ~ "#6d83da95",
      Revised_Source_Niche == "Wild Animal" ~ "#b9553d95"
    ))
  
  # Convert data frame to edge list for igraph
  g <- graph_from_data_frame(d = dists, directed = FALSE)
  
  #Get a list of vertex names
  g_vertex_names <- V(g)$name 
  
  # Match colors to vertices based on names
  vertex_color_vector <- sources$color[match(g_vertex_names, sources$name)]
  
  
  
  #Set a colour for each source
  states <- metadata %>%
    mutate(color = case_when(
      State == "ACT" ~ "blue",
      State == "NSW" ~ "lightskyblue1",
      State == "NT" ~ "darkorange2",
      State == "QLD" ~ "maroon",
      State == "SA" ~ "orangered",
      State == "TAS" ~ "darkgreen",
      State == "VIC" ~ "navy",
      State == "WA" ~ "goldenrod1",
      is.na(State) ~ "grey"
    ))


  #Get a list of vertex names
  g_vertex_names <- V(g)$name 
  
  # Match colors to vertices based on names
  vertex_color_vector_states <- states$color[match(g_vertex_names, states$name)]
  
  
  # Plot the graph
  pdf(file = paste0("../figures/Network_graphs/Network_by_state_",
                    sequence_type,
                    ".pdf"),
      width = 11.69,
      height = 8.27)
  
  E(g)$thickness <- ifelse(E(g)$SNPs <= 20, 3,
                           ifelse(E(g)$SNPs <= 50, 2, 1))
  
  
  set.seed(1)
  
  plot(g, edge.width = E(g)$thickness,
       rescale = TRUE,
       main = sequence_type,
       vertex.size = 3,
       vertex.label = NA,
       vertex.color = vertex_color_vector_states,
       )
  
  legend("bottomleft",
         legend = c("ACT",
             "NSW", 
             "NT", 
             "QLD", 
             "SA", 
             "TAS", 
             "VIC", 
             "WA",
             "Unknown"),
         fill = c(
           "blue",
           "lightskyblue1",
           "darkorange2",
           "maroon",
           "orangered",
           "darkgreen",
           "navy",
           "goldenrod1",
           "grey"),
         pch=19,
         bty = "n",
         pt.cex = 1,
         cex = 1,
         text.col="black" ,
         horiz = FALSE)
  
  dev.off()
}
clusters <- cluster_walktrap(g)

plot(clusters, g, edge.width = E(g)$thickness,
     rescale = TRUE,
     main = sequence_type,
     vertex.size = 3,
     vertex.label = NA,
     mark.groups = T
)



